### PR TITLE
Implement a windowed vertices function in GraphDB in DocROWN

### PR DESCRIPTION
**Pull request by Haaroon**
This implements the following spec 

Right now vertices can only be accessed across the whole graph. As a user, I want to be able to access vertices between a window-frame, see other previous implementations of this: https://pometry.atlassian.net/browse/DEV-1447 and https://pometry.atlassian.net/browse/DEV-1448 

We need to implement a similar-esque function for the the vertices such that they can be retrieved between a certain window.

Note that GraphDB already has a implemented of vertices across time. This ticket doesn’t require any work on the storage side. It only requires a “GETTER” function which retrieves the vertices already stored across time given user parameters. 

Further note that this is already written in [graph.rs](http://graph.rs/) 

Acceptance criteria

Look at  or  to see how they were written

Implement the windowed vertices function which given a start and end time will retrieves vertices that are active within that time (definition of active is loose - so see how the degree and neighbours work)

Write unit tests that confirms the soundness of the function (again see  for the various edge-cases to try)

Document the code according to 

Push and make sure tests pass  

Get code reviewed by  or

### CODE REVIEW

**Comment by Haaroon** [comment]
tests pass locally after merge

File: docbrown/db/src/graphdb.rs [hunk 1]
```diff
@@ -700,6 +713,56 @@ mod db_tests {
         assert_eq!(both_expected, both_actual);
     }
 
+    #[quickcheck]
+    fn vertices_window_test()  {
+        // make 4 graphs with different num of partitions
+        let graph_one = GraphDB::new(1);
+        let graph_three = GraphDB::new(3);
```
Comment (iamsmkr) [inline docbrown/db/src/graphdb.rs:720]: You may probably want to compare results from two different graphs here.

**Review by iamsmkr — CHANGES_REQUESTED** [review]

File: docbrown/db/src/graphdb.rs (hunk 1 above)
Comment (Haaroon) [reply docbrown/db/src/graphdb.rs:720]: ok done

File: docbrown/core/src/graph.rs [hunk 2]
```diff
@@ -564,6 +579,14 @@ impl<'a> VertexView<'a, TemporalGraph> {
     pub fn global_id(&self) -> u64 {
         self.g_id
     }
+    
+    pub fn partition_id(&self) -> usize {
+        self.pid
+    }
+
+    pub fn window(&self) -> Option<Range<i64>> {
+        self.w.clone()
```
Comment (iamsmkr) [inline docbrown/core/src/graph.rs:588]: do we really need to clone this?

File: docbrown/core/src/graph.rs [hunk 3]
```diff
@@ -553,7 +568,7 @@ impl TemporalGraph {
     }
 }
 
-pub(crate) struct VertexView<'a, G> {
+pub(crate) struct VertexView<'a, G: Sized> {
```
Comment (iamsmkr) [inline docbrown/core/src/graph.rs:571]: What purpose does this solve?

File: docbrown/core/src/tpartition.rs [hunk 4]
```diff
@@ -30,6 +31,23 @@ impl<'a> From<EdgeView<'a, TemporalGraph>> for TEdge {
     }
 }
 
+#[derive(Debug)]
+pub struct TVertex {
+    pub g_id: u64, 
+    pid: usize,
```
Comment (iamsmkr) [inline docbrown/core/src/tpartition.rs:37]: TVertex is DTO, if pid is not public probably its not needed here

File: docbrown/db/src/graphdb.rs [hunk 5]
```diff
@@ -144,6 +145,19 @@ impl GraphDB {
         Box::new(iter)
     }
 
+    pub fn vertices_window(
+        &self, 
+        t_start: i64, 
+        t_end: i64,
+    ) -> Box<dyn Iterator<Item = TVertex>+ '_> {
```
Comment (iamsmkr) [inline docbrown/db/src/graphdb.rs:152]: i recommend formatting the code before pushing

**Comment by iamsmkr** [comment]
LGTM

**Comment by Haaroon** [comment]
all tests pass

File: docbrown/core/src/graph.rs (hunk 2 above)
Comment (Haaroon) [reply docbrown/core/src/graph.rs:588]: yup it doesnt implement the copy trait so we have to use clone

**Comment by Haaroon** [comment]
all comments done

**Review by iamsmkr — APPROVED** [review]
